### ISSUE TITLE
Move to proper database

### ISSUE THREAD

**Issue by aseemk** [zoomhub/zoomhub#63]
From our current flat file approach.

The code currently supports using Redis, but not sure if we want to use Redis for our persistent datastore. It's possible; I'm just not sure if it's good.

http://redis.io/topics/persistence
http://oldblog.antirez.com/post/redis-persistence-demystified.html
https://muut.com/blog/technology/redis-as-primary-datastore-wtf.html

@kenperkins and I had a quick chat about this the other day; here was what we discussed:

- Rackspace offers hosted MySQL, which would be convenient from an ops and logistics perspective. But a relational db doesn't feel like the right tool for the job, as our data is very key-value-ish.

- So a key-value store like MongoDB might be nice, but we've heard mixed things about Mongo. Should we look into an alternative? E.g. [RethinkDB](http://rethinkdb.com/) ([see comparison to Mongo](http://rethinkdb.com/docs/rethinkdb-vs-mongodb/)).

- Rackspace acquired ObjectRocket, which offers hosted MongoDB and Redis, so that could be nice. But @kenperkins doesn't think our comped Rackspace billing will cover that.

Whatever we do here, we should also fix the code to use random IDs instead of incremental ones while we're in this space. Incremental IDs are both (a) less scalable and (b) incompatible with our now pre-existing Zoom.it data, which has random IDs. But random IDs means retrying the full transaction (including the "by URL" indexing) in case of collisions, and that impl. depends on the db.

**Comment by aseemk** [comment]
Okay, @kenperkins (re-)discovered that all of our small flat files can cause the system to run out of inodes — which is effectively the same as running out of disk space:

http://www.ivankuznetsov.com/2010/02/no-space-left-on-device-running-out-of-inodes.html

And indeed, we're already pretty much at capacity on our production servers!

```
zoomhub@zoomhub-web-01-iad:~$ df
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/xvda1      20640380 15549864   4042044  80% /
udev              498096        8    498088   1% /dev
tmpfs             101384      252    101132   1% /run
none                5120        0      5120   0% /run/lock
none              506908        0    506908   0% /run/shm

zoomhub@zoomhub-web-01-iad:~$ df -i
Filesystem      Inodes   IUsed  IFree IUse% Mounted on
/dev/xvda1     1310720 1303598   7122  100% /
udev            124524     411 124113    1% /dev
tmpfs           126727     286 126441    1% /run
none            126727       1 126726    1% /run/lock
none            126727       1 126726    1% /run/shm
```

Given that we still have >300k IDs not even in our db (issue #64), and that we want to get rid of our giant in-memory URL-to-ID dictionary (issue #95), this issue is way higher-priority now.

@kenperkins and I agreed that for the short-term, it'll be good enough to just move our flat-files db to Cloud Files. Easier than moving to a whole new datastore; already covered by our comped billing; etc.

This is still a decent chunk of work though. At a high level:

- [ ] Update our code to read to and write from Cloud Files instead of local disk.
- [ ] What about local dev? Separate buckets for separate envs? Or local disk locally?
- [ ] Massive data migration round 2: move all the data from production disk to Cloud Files.

**Comment by aseemk** [comment]
We should version our Cloud Files objects for this db container:

http://www.rackspace.com/blog/rackspace-cloud-files-how-to-use-object-versioning/

**Comment by aseemk** [comment]
With the addition of a proper worker now, and a feature to process old content on-demand (#94), I needed to add proper tracking of content being worked on or not, so this feature needs strong consistency to work effectively.

And come to think of it, we'll also need strong consistency (and/or indexing or uniqueness constraints or something similar) if we want to turn on acceptance of new content (#93), as the same URL can get submitted twice around the same time. (We ran into that at Microsoft, but perhaps it's not a huge deal.)

We know Cloud Files is a stop-gap database, so perhaps for the short-term, we can also solve these problems via an in-memory cache (e.g. [node-lru-cache](https://github.com/isaacs/node-lru-cache)), since we have only one web server.

**Comment by kenperkins** [comment]
We've moved to a block storage volume for a local content buying us a little time:

```bash
root@zoomhub-web-01-iad:/var/data/zoomhub/content-by-id# ls
db.gz  master-database-2014-10-10.gz

root@zoomhub-web-01-iad:/var/data/zoomhub/content-by-id# df -i
Filesystem      Inodes   IUsed   IFree IUse% Mounted on
/dev/xvda1     1310720  117289 1193431    9% /
udev            124524     392  124132    1% /dev
tmpfs           126727     269  126458    1% /run
none            126727       1  126726    1% /run/lock
none            126727       1  126726    1% /run/shm
/dev/xvdb1     6553600 1192883 5360717   19% /mnt/datavolume
```

**Comment by gasi** [comment]
> But a relational db doesn't feel like the right tool for the job, as our data is very key-value-ish.

Interesting POV. I just thought it’d be cool to query our content for the largest sizes and if we stored it all as opaque JSON values, we’d have manually query each entry to get that list, whereas, I believe, SQL would allow us to do that more straightforward.

Based on my research, a million entries should be easily manageable in MySQL or PostgreSQL these days :smile:

**Comment by aseemk** [comment]
Total :+1: at this point to MySQL, if it's hosted for us for free. Agreed we don't need to be "web scale". =)

**Comment by aseemk** [comment]
(FWIW, if we had Mongo or similar, those dbs speak JSON, so you could still query.)

**Comment by gasi** [comment]
Fixed by #120.
